app/src/controllers/reservation/services/ReservationService.ts

```diff
@@ -247,39 +247,92 @@ const ReservationService: ReservationServiceInterface = {
   },
 
   async updateReservation(user, shopId, reservationId, reservationDate, clientId, menuId, stylistId) {
-    if (user.role.slug === RoleSlug.SHOP_STAFF && !await isUserOwnedShop(user.id, shopId)) {
-      Logger.debug('Shop is not owned by user')
-      throw new AuthorizationError()
+    // check if shop exists
+    const shopDetails = await ShopRepository.fetchShopDetailsForReservation(shopId)
+    if (!shopDetails) {
+      Logger.debug('Shop does not exist')
+      throw new NotFoundError()
     }
 
-    const reservationExists = await ReservationRepository.reservationExists(reservationId)
-    if (!reservationExists) {
+    // check if reservation exists
+
+    const currentReservation = await ReservationRepository.fetchShopReservation(shopId, reservationId)
+    if (!currentReservation) {
       Logger.debug('Reservation does not exist')
       throw new NotFoundError()
     }
 
+    // check if new reservation is valid
+
+    if (reservationDate < new Date() || reservationDate === currentReservation.reservationDate) {
+      Logger.debug('Invalid date, earlier than today')
+      throw new NotFoundError()
+    }
+
     const clientExists = await UserRepository.userExists(clientId)
     if (!clientExists) {
       Logger.debug('User does not exist')
       throw new NotFoundError()
     }
 
-    if (!await isValidMenuId(shopId, menuId)) {
-      Logger.debug('Menu does not exist in shop')
+    // check in menu exists
+    const menu = await MenuRepository.fetchShopMenu(shopId, menuId)
+    if (!menu) {
+      Logger.debug('Meny does not exist in shop')
       throw new InvalidParamsError()
     }
 
+    // check shop schedule availability
+    const reservationEndDate = new Date(reservationDate.getTime() + menu.duration * 1000 * 60)
+    const reservationIsWithinShopSchedule = isWithinSchedule(shopDetails.startTime, shopDetails.endTime,
+      shopDetails.days, reservationDate, reservationEndDate, [reservationDate.getDay()])
     if (stylistId && !isValidStylistId(shopId, stylistId)) {
       Logger.debug('Stylist does not exist in shop')
       throw new InvalidParamsError()
     }
 
-    const dateObj = new Date(reservationDate)
-    if (dateObj < today) {
-      Logger.debug('Invalid date, earlier than today')
+    if (!reservationIsWithinShopSchedule) {
+      Logger.debug('Reservation date is not within shop schedule')
       throw new InvalidParamsError()
     }
 
+    // check availability
+    const reservationsForSameDay = await ReservationRepository.fetchShopReservationsForAvailabilityWithMenuDuration(
+      shopId, reservationDate, 1,
+    )
+
+    const conflictingReservations = getConflictingReservations(reservationDate, menu.duration, reservationsForSameDay)
+    if (conflictingReservations.length >= shopDetails.seats) {
+      Logger.debug('Provided time is not available')
+      throw new InvalidParamsError()
+    }
```

Comment:
maybe you should add somewhere in the middle a way to remove the current reservation from the reservations for same day

Reply:
Doesn't the update method remove the update itself?? because its the same reservation being updated?